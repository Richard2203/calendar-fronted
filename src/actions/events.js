import Swal from 'sweetalert2';
import { types } from '../types/types';
import { fetchConToken } from '../helpers/fetch';
import { Events_Parse } from '../helpers/Events_Parse';

export const eventStartAddNew = (event) => {
	return async (dispatch, getState) => {
		const { name, uid } = getState().auth;

		try {
			const resp = await fetchConToken('events', event, 'POST');
			const body = await resp.json();

			if (body.ok) {
				// agregando la propieadad "id" al objeto "event"
				event.id = body.evento.id;

				// agregando la propiedad "user" al objeto "event"
				event.user = { _id: uid, name };

				// guardando el "event" en Redux
				dispatch(eventAddNew(event));
			}
		} catch (error) {
			console.log(error);
		}
	};
};

const eventAddNew = (event) => ({
	type: types.eventAddNew,
	payload: event,
});

export const eventSetActive = (event) => ({
	type: types.eventSetActive,
	payload: event,
});

export const eventClearActiveEvent = () => ({
	type: types.eventClearActiveEvent,
});

export const eventStartUpdated = (event) => {
	return async (dispatch) => {
		try {
			const resp = await fetchConToken(
				`events/${event.id}`,
				event,
				'PUT'
			);
			const body = await resp.json();

			if (body.ok) dispatch(eventUpdated(event));
			else Swal.fire('error', body.msg, 'error');
		} catch (error) {
			console.log(error);
		}
	};
};

const eventUpdated = (event) => ({
	type: types.eventUpdated,
	payload: event,
});

export const eventStartDelete = () => {
	return async (dispatch, getState) => {
		const { id } = getState().calendar.activeEvent;
		try {
			const resp = await fetchConToken(`events/${id}`, {}, 'DELETE');

			const body = await resp.json();

			if (body.ok) dispatch(eventDeleted());
			else Swal.fire('error', body.msg, 'error');
		} catch (error) {
			console.log(error);
		}
	};
};

const eventDeleted = () => ({
	type: types.eventDeleted,
});

export const eventStartLoading = () => {
	return async (dispatch) => {
		try {
			const resp = await fetchConToken('events');
			const { events } = await resp.json();

			// obteniendo los events retornados de la peticion fetch
			const ArrayEvents = Events_Parse(events);

			dispatch(eventLoaded(ArrayEvents));
		} catch (error) {
			console.log(error);
		}
	};
};

const eventLoaded = (events) => ({ type: types.eventLoaded, payload: events });

export const Logout = () => ({ type: types.eventLogout });
