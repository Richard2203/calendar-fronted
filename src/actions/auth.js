import Swal from 'sweetalert2';
import { fetchConToken, fetchSinToken } from '../helpers/fetch';
import { types } from '../types/types';
import { Logout } from './events';

export const startLogin = (userCreden) => {
	return async (dispatch) => {
		const res = await fetchSinToken('auth', userCreden, 'POST');
		const { ok, token, uid, name, msg } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ uid, name }));
		} else Swal.fire('Error', msg, 'error');
	};
};

export const startRegister = (name, email, password) => {
	return async (dispatch) => {
		const res = await fetchSinToken(
			'auth/new',
			{ name, email, password },
			'POST'
		);

		const { ok, name: nam, uid, token, msg } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ nam, uid }));
		} else Swal.fire('Error', msg, 'error');
	};
};

export const startCheking = () => {
	return async (dispatch) => {
		if (!localStorage.getItem('token')) {
			dispatch(checkingFinish());
			return;
		}

		const res = await fetchConToken('auth/renew');

		const { ok, name, uid, token } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ name, uid }));
		} else dispatch(checkingFinish());
	};
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
	return (dispatch) => {
		// implementando el localStorage en un action puesto que de ponerla
		// directamente en el reducer puede marcar error. La logica se debe
		// separar

		// purga todo el localStorage
		localStorage.clear();

		// aqui se purga la informacion en el reducer
		dispatch(authLogout());

		dispatch(Logout());
	};
};

const authLogout = () => ({ type: types.authLogout });

const login = (user) => ({
	type: types.authLogin,
	payload: user,
});
