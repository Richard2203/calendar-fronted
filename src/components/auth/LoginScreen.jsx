import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';

const LoginScreen = () => {
	let dispatch = useDispatch();

	const initialLogin = {
		password: '',
		email: '',
	};

	const [formLogin, handleInputChangeLo] = useForm(initialLogin);

	const { email, password } = formLogin;

	const handleLogin = (e) => {
		e.preventDefault();
		dispatch(startLogin(formLogin));
	};

	const initialRegister = {
		R_name: '',
		R_email: '',
		R_password: '',
		R_passwordCofirm: '',
	};

	const [formRegister, handleInputChangeRe] = useForm(initialRegister);
	const { R_name, R_email, R_password, R_passwordCofirm } = formRegister;

	const handleRegister = (e) => {
		e.preventDefault();
		if (R_password !== R_passwordCofirm)
			Swal.fire('Error', 'Las contrase単as no coinciden', 'error');
		else dispatch(startRegister(R_name, R_email, R_password));
	};

	return (
		<div className="container login-container">
			<div className="row">
				<div className="col-md-6 login-form-1">
					<h3>Ingreso</h3>
					<form onSubmit={handleLogin}>
						<div className="form-group">
							<input
								type="email"
								className="form-control"
								placeholder="Correo"
								name="email"
								value={email}
								onChange={handleInputChangeLo}
							/>
						</div>

						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Contrase単a"
								name="password"
								value={password}
								onChange={handleInputChangeLo}
							/>
						</div>

						<div className="form-group">
							<input
								type="submit"
								className="btnSubmit"
								value="Login"
							/>
						</div>
					</form>
				</div>

				<div className="col-md-6 login-form-2">
					<h3>Registro</h3>
					<form onSubmit={handleRegister}>
						<div className="form-group">
							<input
								type="text"
								className="form-control"
								placeholder="Nombre"
								value={R_name}
								onChange={handleInputChangeRe}
								name="R_name"
							/>
						</div>
						<div className="form-group">
							<input
								type="email"
								className="form-control"
								placeholder="Correo"
								value={R_email}
								onChange={handleInputChangeRe}
								name="R_email"
							/>
						</div>
						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Contrase単a"
								value={R_password}
								onChange={handleInputChangeRe}
								name="R_password"
							/>
						</div>

						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Repita la contrase単a"
								value={R_passwordCofirm}
								onChange={handleInputChangeRe}
								name="R_passwordCofirm"
							/>
						</div>

						<div className="form-group">
							<input
								type="submit"
								className="btnSubmit"
								value="Crear cuenta"
							/>
						</div>
					</form>
				</div>
			</div>
		</div>
	);
};
export default LoginScreen;
