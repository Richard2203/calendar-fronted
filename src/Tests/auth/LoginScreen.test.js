import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import LoginScreen from '../../components/auth/LoginScreen';
import { act } from '@testing-library/react';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

jest.mock('sweetalert2', () => ({
	fire: jest.fn(),
}));

jest.mock('../../actions/auth', () => ({
	startRegister: jest.fn(),
	startLogin: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas en <LoginScreen/>', () => {
	const wrapper = mount(
		<Provider store={store}>
			<LoginScreen />
		</Provider>
	);

	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('debe renderizar correctamente el componente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe llamar startLogin con los argumentos correspondientes', () => {
		act(() => {
			wrapper.find('input[name="email"]').prop('onChange')({
				target: {
					name: 'email',
					value: 'paquita@gmail.com',
				},
			});
		});

		act(() => {
			wrapper.find('input[name="password"]').prop('onChange')({
				target: {
					name: 'password',
					value: '1234567890',
				},
			});
		});

		act(() => {
			wrapper.find('form').at(0).prop('onSubmit')({
				preventDefault() {},
			});
		});

		expect(startLogin).toHaveBeenCalledWith(expect.any(Object));
	});

	test('debe llamar startRegister con los argumentos correspondientes', () => {
		wrapper.find('input[name="R_name"]').simulate('change', {
			target: {
				name: 'R_name',
				value: 'paquita',
			},
		});

		wrapper.find('input[name="R_email"]').simulate('change', {
			target: {
				name: 'R_email',
				value: 'paquita@gmail.com',
			},
		});

		wrapper.find('input[name="R_password"]').simulate('change', {
			target: {
				name: 'R_password',
				value: '1234567890',
			},
		});

		wrapper.find('input[name="R_passwordC"]').simulate('change', {
			target: {
				name: 'R_passwordC',
				value: '1234567890',
			},
		});

		wrapper.find('form').at(1).prop('onSubmit')({
			preventDefault() {},
		});

		expect(startRegister).toHaveBeenCalledWith(
			'paquita',
			'paquita@gmail.com',
			'1234567890'
		);
	});

	test('debe llamar a fire con los argumentos correspondientes', () => {
		wrapper.find('input[name="R_name"]').simulate('change', {
			target: {
				name: 'R_name',
				value: 'paquita',
			},
		});

		wrapper.find('input[name="R_email"]').simulate('change', {
			target: {
				name: 'R_email',
				value: 'paquita@gmail.com',
			},
		});

		wrapper.find('input[name="R_password"]').simulate('change', {
			target: {
				name: 'R_password',
				value: '1234567890',
			},
		});

		wrapper.find('input[name="R_passwordC"]').simulate('change', {
			target: {
				name: 'R_passwordC',
				value: '12340',
			},
		});

		wrapper.find('form').at(1).prop('onSubmit')({ preventDefault() {} });

		expect(Swal.fire).toHaveBeenCalled();
		expect(Swal.fire).toHaveBeenCalledWith(
			'Error',
			'Las contrase√±as no coinciden',
			'error'
		);
	});
});
