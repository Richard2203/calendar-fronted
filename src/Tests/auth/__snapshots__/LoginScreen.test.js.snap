// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen/> debe renderizar correctamente el componente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        id="formLogin"
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="R_name"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="R_email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="R_password"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="R_passwordC"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
